src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs

```diff
@@ -182,21 +185,45 @@ public bool HandlePartialDiscoveryComplete(IProxyDiscoveryManager proxyDiscovery
 
     private List<ProviderSpecificWorkload<DiscoveryCriteria>> SplitToWorkloads(DiscoveryCriteria discoveryCriteria, Dictionary<string, TestRuntimeProviderInfo> sourceToTestHostProviderMap)
     {
+        var sources = discoveryCriteria.Sources;
+        // Each source is grouped with its respective provider.
+        var providerGroups = sources
+            .Select(source => new ProviderSpecificWorkload<string>(source, sourceToTestHostProviderMap[source]))
+            .GroupBy(psw => psw.Provider);
+
         List<ProviderSpecificWorkload<DiscoveryCriteria>> workloads = new();
-        foreach (var source in discoveryCriteria.Sources)
+        foreach (var group in providerGroups)
         {
-            var testHostProviderInfo = sourceToTestHostProviderMap[source];
-            var runsettingsXml = testHostProviderInfo.RunSettings;
-            var updatedDiscoveryCriteria = new ProviderSpecificWorkload<DiscoveryCriteria>(NewDiscoveryCriteriaFromSourceAndSettings(source, discoveryCriteria, runsettingsXml), testHostProviderInfo);
-            workloads.Add(updatedDiscoveryCriteria);
+            var testhostProviderInfo = group.Key;
+            // If the run is not parallel and the host is shared, put all testcases on single testhost.
```

Comment:
Might be worth to explain briefly why we don't think it adds value to share for parallel.

Reply:
done